### PR TITLE
feat: delete kube context

### LINKED ISSUE #4417: Kube context delete
### Is your enhancement related to a problem? Please describe

As Kube contexts are going to be first class citizen in Podman Desktop and displayed nicely, we should be able to delete them as Kubeconfig may have a lot of obsoleted contexts

### Describe the solution you'd like

On the Kubecontext UI we should be able to see a delete button/icon

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by feloy**
Signed-off-by: Philippe Martin <phmartin@redhat.com>

#### Settings > Kubernetes

This PR adds a Trash button for each item in the list. 
Clicking the button deletes the context from the list and from the kubeconfig file.

/!\ Any user or cluster which were referenced by the context and not referenced anymore in the kubeconfig file is also removed from the file.

https://github.com/containers/podman-desktop/assets/9973512/7f90f800-2791-4d62-9732-6b0fe0e8ae5f

TODO
- [ ] do we want to ask for confirmation before to delete the context from the file?
- [ ] how do we want to display an error when one occurs (for example if the kubeconfig file is not writable)?
- [ ] do we want to update the current-context?
- [ ] unit tests

### What issues does this PR fix or reference?

Fixes #4417 

### How to test this PR?

First make a backup copy of your kubeconfig file, as it will be modified by the test.

Press the Trash button on any of the entries.

### CODE REVIEW

**Comment by afbjorklund** [comment]
After deleting the kubernetes context, the current-context still references it.

```
apiVersion: v1
kind: Config
clusters: []
users: []
contexts: []
preferences: {}
current-context: kubernetes-admin@kubernetes
```

After that, `kubectl` starts throwing errors: (and you can't really "deselect" it, in GUI)

`invalid configuration: [context was not found for specified context: kubernetes-admin@kubernetes, cluster has no server defined]`


Have to use the terminal, to fix it:

```
kubectl config unset current-context
```

But `delete-context` does the same thing: (only difference is that it warns you that it got broken)

```
kubectl config delete-context kubernetes-admin@kubernetes
```

`warning: this removed your active context, use "kubectl config use-context" to select a different one`

**Comment by feloy** [comment]
Based on @deboer-tim and @afbjorklund feedback, we could block the deletion for the current context (but we will need to implement before the change of current context from the GUI), except when this is the latest context in the file: in this case we would unet the current context.

File: packages/main/src/plugin/kubernetes-client.ts
Comment (jeffmaury) [inline packages/main/src/plugin/kubernetes-client.ts:285]: What if the current context if deleted

File: packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte
Comment (jeffmaury) [inline packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte:45]: That should not be necessary once store is there, don't know how to handle it once it's been merged

**Comment by benoitf** [comment]
hello, I'm wondering why I could not remove the current context ?

let say my context is an old kind cluster that I have deleted, why should not be able to get rid of it ?

**Comment by benoitf** [comment]
rebased to include https://github.com/containers/podman-desktop/pull/4807

**Comment by deboer-tim** [comment]
> hello, I'm wondering why I could not remove the current context ?
> 
> let say my context is an old kind cluster that I have deleted, why should not be able to get rid of it ?

We talked about #4417 on a call yesterday but maybe we weren't thinking about some cases. My recommendation would be a confirmation dialog if you're deleting the current context, but no confirmation dialog otherwise. (i.e. you can delete any context, you just get prompted before current because it's 'in use'/special, and I'm assuming we would unset it).

We don't have the ability to see or change the current context yet (issue #4618 later this sprint) so this part is a little cart-horse.

**Comment by feloy** [comment]
I have made the changes according to the recommendations from @deboer-tim, by using a confirmation dialog before to delete the current context.

File: packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte
Comment (jeffmaury) [inline packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte]:
```suggestion
    if (result?.response !== 0) {
```

Comment (jeffmaury) [reply packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte]: Not sure why we check for !result given the method signature

**Review by jeffmaury — COMMENTED** [review]
Maybe we should have a unit test on the kubernetes client ?

File: packages/main/src/plugin/kubernetes-client.ts
Comment (feloy) [reply packages/main/src/plugin/kubernetes-client.ts:285]: This is my question above (`do we want to update the current-context?`). 

As it is now, it is the same behaviour as kubectl:
```
$ kubectl config current-context 
my-current-context
$ kubectl config delete-context my-current-context
warning: this removed your active context, use "kubectl config use-context" to select a different one
deleted context my-current-context from /home/phmartin/.kube/config
$ kubectl config current-context 
my-current-context
```

Do we want to keep this same behaviour, or have a different one?

File: packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte
Comment (feloy) [reply packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte:45]: Yes, this will change once #4745 is merged. I would prefer to wait for #4745 to be merged before to continue on this PR, so I can handle this accordingly.

File: packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte
Comment (feloy) [reply packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte]: yes, right, it should not be necessary. Thanks

File: packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte
Comment (jeffmaury) [reply packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte:45]: I think that would be the rebase of #4745 to remove what needs to be removed and this PR should not be blocked

File: packages/main/src/plugin/kubernetes-client.ts
Comment (jeffmaury) [reply packages/main/src/plugin/kubernetes-client.ts:285]: Ignore that was kubectl behavior, thus this is ok for me

**Review by jeffmaury — APPROVED** [review]
LGTM

File: packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte
Comment (feloy) [reply packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte:45]: When using the store, I think we won't have to get the contexts as result of ` window.kubernetesDeleteContext` (as it will come from the store), but we will only need to catch the error. But we will probably need to have a separate variable holding the errors, something like:

```
try {
    await window.kubernetesDeleteContext(contextName);
  } catch (e: unknown) {
    if (e instanceof Error) {
+      errors[contextName] = e;
-      contexts = setKubeUIContextError(contexts, contextName, e);
    }
  }
```

**Review by deboer-tim — APPROVED** [review]
I guess I hadn't manually cleaned up my contexts for a month, this was incredibly useful and far easier/faster than command line or editing. :)

Confirmation before deleting my current context worked well.

**Pull request merged**
